src/theory/strings/regexp_elim.cpp

```diff
@@ -0,0 +1,441 @@
+      {
+        // otherwise, we can use indexof to represent some next occurrence
+        if (gap_exact[i + 1] && i + 1 != size)
+        {
+          if (!options::regExpElimAgg())
+          {
+            success = false;
+            break;
+          }
+          // if the gap after this one is strict, we need a non-greedy find
+          // thus, we add a symbolic constant
+          Node k = nm->mkBoundVar(nm->integerType());
+          non_greedy_find_vars.push_back(k);
+          prev_end = nm->mkNode(PLUS, prev_end, k);
+        }
+        Node curr = nm->mkNode(STRING_STRIDOF, x, sc, prev_end);
+        Node idofFind = curr.eqNode(d_neg_one).negate();
+        conj.push_back(idofFind);
+        prev_end = nm->mkNode(PLUS, curr, lensc);
+      }
+      // if applicable, process the last gap
+      if (i == (size - 1))
+      {
+        if (gap_exact_end)
+        {
+          // substring relative to the end
+          Node curr = nm->mkNode(MINUS, lenx, lensc);
+          if (gap_minsize_end > 0)
+          {
+            curr =
+                nm->mkNode(MINUS, curr, nm->mkConst(Rational(gap_minsize_end)));
+          }
+          Node ss = nm->mkNode(STRING_SUBSTR, x, curr, lensc);
+          conj.push_back(ss.eqNode(sc));
+        }
+        else if (gap_minsize_end > 0)
+        {
+          Node fit = nm->mkNode(
+              LEQ,
+              nm->mkNode(
+                  PLUS, prev_end, nm->mkConst(Rational(gap_minsize_end))),
+              lenx);
+          conj.push_back(fit);
+        }
+      }
+    }
+    if (success)
+    {
+      // since sep_children is non-empty, conj is non-empty
+      Assert(!conj.empty());
+      Node res = conj.size() == 1 ? conj[0] : nm->mkNode(AND, conj);
+      // process the non-greedy find variables
+      if (!non_greedy_find_vars.empty())
+      {
+        std::vector<Node> children;
+        for (const Node& v : non_greedy_find_vars)
+        {
+          Node bound = nm->mkNode(
+              AND, nm->mkNode(LEQ, d_zero, v), nm->mkNode(LT, v, lenx));
+          children.push_back(bound);
+        }
+        children.push_back(res);
+        Node body = nm->mkNode(AND, children);
+        Node bvl = nm->mkNode(BOUND_VAR_LIST, non_greedy_find_vars);
+        res = nm->mkNode(EXISTS, bvl, body);
+      }
+      // e.g., writing "A" for (str.to.re "A") and _ for re.allchar:
+      //   x in (re.++ "A" (re.* _) "B" (re.* _)) --->
+      //     substr(x,0,1)="A" ^ indexof(x,"B",1)!=-1
+      //   x in (re.++ (re.* _) "A" _ _ _ (re.* _) "B" _ _ (re.* _)) --->
+      //     indexof(x,"A",0)!=-1 ^
+      //     indexof( x, "B", indexof( x, "A", 0 ) + 1 + 3 ) != -1 ^
+      //     indexof( x, "B", indexof( x, "A", 0 ) + 1 + 3 )+1+2 <= len(x)
+      return returnElim(atom, res, "concat-with-gaps");
+    }
+  }
+
+  if (!options::regExpElimAgg())
+  {
+    return Node::null();
+  }
+  // only aggressive rewrites below here
+
+  // if the first or last child is constant string, split
```

Comment:
A better description than "split" would be helpful here. Also, why is this an aggressive rewrite? This one seems quite harmless.

Reply:
The reason this could be bad is that confuses the regexp solver.  Say x in R1 and x in R2, the regexp solver has the chance to show unsat by showing R1 intersect R2 is empty. If we change x in R1 to substr( x, n1, n2 ) in R1' ^ ..., then we don't.

Reply:
Thanks for the explanation, that makes sense.